Upgrade React from 15 to 16

magopian opened the pull request:
Fixes #564

- [X] Update package versions
- [X] Migrate old react-router's IndexRoute and IndexRedirect
- [x] Migrate "common" components route (Notifications, Sidebar)
- [X] Migrate HomePage route (including AuthForm)
- [x] Migrate plugins dynamic routes
- [x] Migrate breadcrumbs
- [x] Migrate Buckets route
- [x] Migrate groups route
- [x] Migrate Collections route
- [x] Migrate Records route
- [x] Migrate 404 route
- [x] `location.query` [doesn't exist anymore in react-router v4](https://codeburst.io/react-router-v4-unofficial-migration-guide-5a370b8905a#27d7)
- [x] Update the react-redux integration by replacing react-router-redux with [connected-react-router](https://github.com/supasate/connected-react-router)
- [x] There's no `onEnter` nor `onUpdate` on `Route`s [anymore](https://codeburst.io/react-router-v4-unofficial-migration-guide-5a370b8905a#e85f)
- [x] Use a `SmartRoute` HOC (see [CrumbRoute](https://github.com/svenanders/react-breadcrumbs/blob/b71986a33284188bd3f83d47c34eb8a60c00d35b/demo/src/crumb-route.jsx#L16)) that
  - [x] calls `routeActions.routeUpdated(params, location)` with the `Route` params and location
  - [x] adds a `Breadcrumb` component (to fix the breadcrumbs)
- [x] fix the `Sidebar` so it properly highlights the current bucket and collection when needed
- [x] fix tests
- [x] fix flow
- [x] fix issue with [bucket history](http://localhost:3000/#/buckets/testbucket/history) and [group history](http://localhost:3000/#/buckets/testbucket/groups/testgroup/history) and [collections history](http://localhost:3000/#/buckets/testbucket/collections/testcollection/history) and [records history](http://localhost:3000/#/buckets/testbucket/collections/testcollection/records/<:rid>/history) when refreshing the page (`TypeError: routing.locationBeforeTransitions is null`)
- [x] fix issue with records being listed several times in the collection's record list

leplatrem commented:
I tested locally with the signoff plugin etc. and it works great :)

leplatrem commented on src/components/AdminLink.js:
```diff
@@ -4,21 +4,26 @@ import type { RouteParams } from "../types";
     const { children, name, params, ...linkProps } = this.props;
+    const toUrl = url(name, params);
+    const to =
+      (linkProps.query && `${toUrl}?${stringify(linkProps.query)}`) || toUrl;
```
``const to = linkProps.query ? `${toUrl}?${stringify(linkProps.query)}` : toUrl``

magopian replied:
Is it more readable? It's fine with me, i'll change it accordingly ;)

leplatrem commented on src/components/AdminLink.js:
```diff
@@ -4,21 +4,26 @@ import type { RouteParams } from "../types";
+      (linkProps.query && `${toUrl}?${stringify(linkProps.query)}`) || toUrl;
     return (
-      <Link {...linkProps} to={url(name, params)}>
+      <Link {...linkProps} to={to}>
```
Couldn't we have written ` linkProps = { ...linkProps, to}` above and just pass `{ ...linkProps }` here?

magopian replied:
We could have, but why? What for? Is it more readable? In that case, why not

```<Link {...linkProps, to}>```?

leplatrem replied:
`<Link {...linkProps, to}>` is nice. Not super important though.

leplatrem commented on src/components/bucket/BucketAttributes.js:
```diff
@@ -35,14 +35,14 @@ export default class BucketAttributes extends PureComponent<Props> {
-    const { params, updateBucket } = this.props;
-    const { bid } = params;
+    const { match, updateBucket } = this.props;
+    const { bid } = match.params;
```
`const { params: { bid } } = match`

leplatrem commented on src/components/bucket/BucketAttributes.js:
```diff
@@ -35,14 +35,14 @@ export default class BucketAttributes extends PureComponent<Props> {
-    const { params, session, capabilities, bucket } = this.props;
-    const { bid } = params;
+    const { match, session, capabilities, bucket } = this.props;
+    const { bid } = match.params;
```
`const { params: { bid } } = match`

leplatrem commented on src/components/collection/CollectionHistory.js:
```diff
@@ -18,23 +20,47 @@ type Props = {
 
+export const onCollectionHistoryEnter = (props: Props) => {
+  const { listCollectionHistory, match, session, location } = props;
+  const { bid, cid } = match.params;
```
`const {params: { bid, cid }} = match;`

leplatrem commented on src/components/group/GroupAttributes.js:
```diff
@@ -26,22 +26,22 @@ type Props = {
-    const { params, updateGroup } = this.props;
-    const { bid, gid } = params;
+    const { match, updateGroup } = this.props;
+    const { bid, gid } = match.params;
```
`const { params: { bid, gid }} = match;`

leplatrem commented on src/index.js:
```diff
@@ -56,16 +54,11 @@ export default class KintoAdmin extends Component<Props> {
-          {getRoutes(this.store, registerPlugins)}
-        </Router>
+        <ConnectedRouter history={hashHistory}>
+          <Switch>{getRoutes(this.store, registerPlugins)}</Switch>
```
Shouldn't the `<Switch>` be in `getRoutes()`? (genuine question, maybe not)

magopian commented:
@leplatrem I believe I've addressed your comments, but the [one about `linkProps`](https://github.com/Kinto/kinto-admin/pull/567#discussion_r204450727) that I might not have understood.

Please let me know what you think!

leplatrem commented:
> I believe I've addressed your comments, 

I still see those `const { bid, cid } = match.params;`. I think we should all write them the same way :)

magopian commented:
> I still see those const { bid, cid } = match.params;. I think we should all write them the same way :)

Ah, of course, I only fixed the ones you pointed out, sorry about that, fix incoming :P

leplatrem commented:
(please do not merge without squashing)
